java-viewer/src/main/resources/public/lib/viewer/viewer.js

```diff
@@ -0,0 +1,465 @@
+  });
+  result += '</table>';
+  return result;
+}
+
+function setNodesColor(nodes, selectedNodesIds, forcedHighlighting) {
+  nodes.forEach(function (node) {
+    // common properties
+    node['color'] = {
+      background: '#eee',
+      border: 'gray',
+      highlight: {
+        background: 'yellow',
+        border: 'gold'
+      }
+    };
+    node['font'] = {
+      size: 12,
+      face: 'monospace',
+      color: '#333',
+      align: 'left'
+    };
+
+    const highlighting = $.inArray(node.id, selectedNodesIds) !== -1 ? forcedHighlighting : node.highlighting;
+
+    let newBackgroundColor, newBorderColor, newFontColor;
+    switch(highlighting) {
+      case 'firstNode':
+        newBackgroundColor = 'palegreen';
+        newBorderColor = 'limegreen';
+        break;
+      case 'exitNode':
+        newBackgroundColor = 'black';
+        newBorderColor = 'dimgray';
+        newFontColor = 'white';
+        break;
+      case 'lostNode':
+        newBackgroundColor = 'red';
+        newBorderColor = 'firebrick';
+        newFontColor = 'white';
+        break;
+      case 'tokenKind':
+        newBackgroundColor = 'black';
+        newBorderColor = 'dimgray';
+        newFontColor = 'white';
+        node['shape'] = 'box';
+        break;
+      case 'classKind':
+        newBackgroundColor = 'pink';
+        newBorderColor = 'red';
+        break;
+      case 'methodKind':
+        newBackgroundColor = 'skyblue';
+        newBorderColor = 'blue';
+        break;
+      case 'samePP':
+        newBackgroundColor = 'pink';
+        newBorderColor = 'mediumvioletred';
+        newFontColor = 'black';
+        break;
+    }
+    if (newBackgroundColor) {
+      node['color']['background'] = newBackgroundColor;
+    }
+    if (newBorderColor) {
+      node['color']['border'] = newBorderColor;
+    }
+    if (newFontColor) {
+      node['font']['color'] = newFontColor;
+    }
+    nodes.update(node);
+  });
+}
+
+function setEdgesColor(edges) {
+  edges.forEach(function (edge) {
+    // common properties
+    edge['color'] = {
+      color: 'gray',
+      highlight: 'yellow'
+    };
+    edge['font'] = {
+      size: 10,
+      color: 'gray'
+    };
+
+    switch(edge.highlighting) {
+      case 'exceptionEdge':
+        edge['color']['color'] = 'red';
+        edge['font']['color'] = 'red';
+        break;
+      case 'yieldEdge':
+        edge['color']['color'] = 'purple';
+        edge['font']['color'] = 'purple';
+        break;
+      default:
+        // do nothing
+    }
+
+    edges.update(edge);
+  });
+}
+
+function getPPMap(nodes) {
+  let result = {};
+  nodes.forEach(function (node) {
+    const details = node['details'];
+    if (details.ppKey) {
+      if (!result[details.ppKey]) {
+        result[details.ppKey] = [];
+      }
+      result[details.ppKey].push(node.id);
+    }
+  });
+  return result;
+}
+
+function mapPPByLine(cfgCode) {
+  let result = {};
+  let currentBlock;
```

Comment:
I need `currentBlock` to be outside the `for` loop as I'll compute it when encountered on certain lines, and then use it for next ones.